marimo/_server/api/endpoints/xterm.py

```diff
@@ -0,0 +1,82 @@
+from __future__ import annotations
+
+import asyncio
+import os
+import pty
+import signal
+import subprocess
+
+from starlette.websockets import WebSocket, WebSocketDisconnect
+
+from marimo import _loggers
+from marimo._server.router import APIRouter
+
+LOGGER = _loggers.marimo_logger()
+
+router = APIRouter()
+
+
+async def _read_from_pty(master: int, websocket: WebSocket) -> None:
+    with os.fdopen(master, "rb", buffering=0) as master_file:
+        while True:
+            data = await asyncio.get_event_loop().run_in_executor(
+                None, master_file.read, 1024
+            )
```

Comment:
i dont know python well enough to cancel this. it always hangs on this when the server dies